### ISSUE TITLE
excessive RAM usage with quantized base model and LORA with SFTTrainer

### ISSUE THREAD

**Issue by galtay** [huggingface/trl#1489]
installed trl from git clone of main

git rev-parse main
6c2f829bb7408660b0e581cde56fbff0980b9d7b

- `transformers` version: 4.39.1
- Platform: Linux-6.5.0-26-generic-x86_64-with-glibc2.35
- Python version: 3.11.8
- Huggingface_hub version: 0.22.0
- Safetensors version: 0.4.2
- Accelerate version: 0.28.0
- peft version: 0.10.0

The following command is using 21.4GB of RAM on an RTX3090. 
trainable params: 7,127,040 || all params: 338,323,456 || trainable%: 2.106575785274551

with 4bit quantization i was expecting the RAM usage to be much less. am i missing something? 
(only change I made was #1488 to be able to use `all-linear` for `lora_target_modules` 

```shell
CUDA_VISIBLE_DEVICES=0 TRL_USE_RICH=1 python examples/scripts/sft.py \
       --model_name_or_path="facebook/opt-350m" \
       --report_to="wandb" \
       --learning_rate=1.41e-5 \
       --per_device_train_batch_size=32 \
       --gradient_accumulation_steps=16 \
       --output_dir="sft_openassistant-guanaco" \
       --logging_steps=1 \
       --num_train_epochs=3 \
       --max_steps=-1 \
       --gradient_checkpointing \
       --use_peft \
       --lora_r=16 \
       --lora_alpha=32 \
       --lora_target_modules="all-linear" \
       --load_in_4bit \
       --bnb_4bit_quant_type="nf4" \
       --use_bnb_nested_quant \
       --attn_implementation="flash_attention_2" \
       --torch_dtype="float16" \
       --max_seq_length=512
```

wandb link https://wandb.ai/galtay/huggingface/runs/x99vv79m

**Comment by galtay** [comment]
@lewtun i might be totally misunderstanding what this example script does, but I was expecting that fine tuning a 4bit  quantized 350m parameter model with lora would not require much more RAM than the base model in eval mode.

**Comment by galtay** [comment]
adding gradient checkpointing brings it down to 13GB of GPU RAM, but that still seems like an awful lot ... like something is going wrong with the quantization or the base model parameters are holding gradients or optimizer state 

```shell
CUDA_VISIBLE_DEVICES=0 TRL_USE_RICH=1 python examples/scripts/sft.py \
       --model_name_or_path="facebook/opt-350m" \
       --report_to="wandb" \
       --learning_rate=1.41e-5 \
       --per_device_train_batch_size=32 \
       --gradient_accumulation_steps=16 \
       --output_dir="sft_openassistant-guanaco" \
       --logging_steps=1 \
       --num_train_epochs=3 \
       --max_steps=-1 \
       --gradient_checkpointing=True \
       --gradient_checkpointing_use_reentrant=True \
       --use_peft \
       --lora_r=16 \
       --lora_alpha=32 \
       --lora_target_modules="all-linear" \
       --load_in_4bit \
       --bnb_4bit_quant_type="nf4" \
       --use_bnb_nested_quant \
       --attn_implementation="flash_attention_2" \
       --torch_dtype="float16" \
       --max_seq_length=512
```

**Comment by galtay** [comment]
hey @younesbelkada, thank you for taking a look! 

yea, I forgot to consider the large-ish amount of RAM in the logits ... I guess (32 batch) * (50272 vocab) * (512 seq len) * (16 bits) / 2**20 ~ 12.5GB by itself ... I'll close this for now and try and do a finer grained accounting of where the RAM is actually going ... and maybe try some larger models and smaller batch sizes ;)
